### PR TITLE
add log viewer CLI

### PR DESCRIPTION
**Pull request by nbren12**
This adds a small webapp for helping to view the GFDL logs from a
segmented run. Currently, this only plots the globally averaged
variables.

Added public API:
- CLI: `prognostic_run_diags logs`

Requirement changes:
- Install streamlit (see license information below).

Screenshot: 
Screen Shot 2021-05-19 at 1 42 22 PM

### CODE REVIEW

File: constraints.txt [hunk 1]
```diff
@@ -15,18 +15,21 @@ appnope==0.1.2
 argcomplete==1.12.0
 argon2-cffi==20.1.0
 asciitree==0.3.3
+astor==0.8.1
```
Comment (nbren12) [inline constraints.txt:18]: License: BSD-3-CLAUSE

File: constraints.txt [hunk 2]
```diff
@@ -15,18 +15,21 @@ appnope==0.1.2
...
 astunparse==1.6.3
 async-generator==1.10
 async-timeout==3.0.1
 attrs==19.3.0
 avro-python3==1.9.2.1
 babel==2.8.0
 backcall==0.2.0
+base58==2.1.0
```
Comment (nbren12) [inline constraints.txt:26]: MIT

File: constraints.txt [hunk 3]
```diff
@@ -15,18 +15,21 @@ appnope==0.1.2
...
 bleach==3.3.0
+blinker==1.4
```
Comment (nbren12) [inline constraints.txt:28]: MIT

File: constraints.txt [hunk 4]
```diff
@@ -62,6 +65,8 @@ future==0.18.2
 fv3config==0.8.0
 gast==0.3.3
 gcsfs==2021.4.0
+gitdb==4.0.7
```
Comment (nbren12) [inline constraints.txt:68]: BSD

File: constraints.txt [hunk 5]
```diff
@@ -62,6 +65,8 @@ future==0.18.2
...
+gitpython==3.1.17
```
Comment (nbren12) [inline constraints.txt:69]: BSD

File: constraints.txt [hunk 6]
```diff
@@ -101,6 +106,7 @@ intake==0.6.0
 ipykernel==5.5.0
 ipython-genutils==0.2.0
 ipython==7.16.1
+ipywidgets==7.6.3
```
Comment (nbren12) [inline constraints.txt:109]: BSD

File: constraints.txt [hunk 7]
```diff
@@ -112,6 +118,7 @@ jupyter-packaging==0.7.12
 jupyter-server==1.4.1
 jupyterlab-pygments==0.1.2
 jupyterlab-server==2.3.0
+jupyterlab-widgets==1.0.0
```
Comment (nbren12) [inline constraints.txt:121]: BSD License (BSD-3-Clause)

File: constraints.txt [hunk 8]
```diff
@@ -179,6 +186,7 @@ pyasn1-modules==0.2.7
 pyasn1==0.4.8
 pycparser==2.20
 pyct==0.4.8
+pydeck==0.6.2
```
Comment (nbren12) [inline constraints.txt:189]: Apache 2.0 License

File: constraints.txt [hunk 9]
```diff
@@ -202,6 +210,7 @@ scipy==1.5.4
 send2trash==1.5.0
 shellingham==1.3.2
 six==1.15.0
+smmap==4.0.0
```
Comment (nbren12) [inline constraints.txt:213]: BSD

File: constraints.txt [hunk 10]
```diff
@@ -214,6 +223,7 @@ sphinxcontrib-htmlhelp==1.0.3
 sphinxcontrib-jsmath==1.0.1
 sphinxcontrib-qthelp==1.0.3
 sphinxcontrib-serializinghtml==1.1.4
+streamlit==0.82.0
```
Comment (nbren12) [inline constraints.txt:226]: Apache 2

File: constraints.txt [hunk 11]
```diff
@@ -232,13 +242,16 @@ traitlets==4.3.3
 typeguard==2.10.0
 typing-extensions==3.7.4.3
 typing-inspect==0.6.0
+tzlocal==2.1
```
Comment (nbren12) [inline constraints.txt:245]: MIT

File: constraints.txt [hunk 12]
```diff
@@ -232,13 +242,16 @@ traitlets==4.3.3
...
 urllib3==1.25.10
+validators==0.18.2
```
Comment (nbren12) [inline constraints.txt:247]: MIT

File: constraints.txt [hunk 13]
```diff
@@ -232,13 +242,16 @@ traitlets==4.3.3
...
 virtualenv==20.0.33
 wcwidth==0.2.5
 webencodings==0.5.1
 websocket-client==0.57.0
 werkzeug==1.0.1
 wheel==0.35.1
+widgetsnbextension==3.5.1
```
Comment (nbren12) [inline constraints.txt:254]: BSD

**Review by nbren12 — COMMENTED** [review]
All the added dependencies have licenses on our per-approved list.

File: constraints.txt [hunk 14]
```diff
@@ -164,6 +171,7 @@ pint==0.15
 pip-tools==5.3.1
 pkgconfig==1.5.1
 pkginfo==1.5.0.1
+plotly==4.14.3
```
Comment (nbren12) [inline constraints.txt:174]: MIT

File: constraints.txt [hunk 15]
```diff
@@ -196,12 +205,14 @@ recommonmark==0.7.1
 requests-oauthlib==1.3.0
 requests-toolbelt==0.8.0
 requests==2.24.0
+retrying==1.3.3
```
Comment (nbren12) [inline constraints.txt:208]: Apache Software License (Apache 2.0)

File: workflows/prognostic_run_diags/fv3net/diagnostics/prognostic_run/apps/log_viewer.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    df = pd.melt(df.reset_index(), id_vars=["index"])
+    fig = px.line(df, x="index", y="value", color="variable", title=url).update_layout(
+        font=dict(size=10)
+    )
+    return st.plotly_chart(fig)
+
+
+def register_parser(subparsers):
+    parser = subparsers.add_parser(
+        "log-viewer", help="Webapp for viewing segmented run logs."
```
Comment (oliverwm1) [inline workflows/prognostic_run_diags/fv3net/diagnostics/prognostic_run/apps/log_viewer.py:54]: to me, a 'log-viewer' is a tool to view or query logs, not make plots. IMO we'll probably want to extend this to the other text outputs, so maybe we should call it `statistic-viewer` or something? Or be a bit more specific in the help message.

**Review by oliverwm1 — COMMENTED** [review]
Generally looks good! I have a few small comments.

File: workflows/prognostic_run_diags/fv3net/diagnostics/prognostic_run/apps/log_viewer.py (hunk 16 above)
Comment (nbren12) [reply workflows/prognostic_run_diags/fv3net/diagnostics/prognostic_run/apps/log_viewer.py:54]: Logs is a somewhat established convention here. These statistics are parsed from the logs file using the functionality in `vcm.fv3.logs`. I will add some information to help.

Comment (nbren12) [reply workflows/prognostic_run_diags/fv3net/diagnostics/prognostic_run/apps/log_viewer.py]: Also "view" is where all the plotting code is in this prognostic run.

**Review by oliverwm1 — APPROVED** [review]
LGTM

**Pull request merged**